optimum/gptq/quantizer.py

```diff
@@ -0,0 +1,571 @@
+
+        Returns:
+            `nn.Module`: The quantized model
+        """
+
+        if not torch.cuda.is_available():
+            raise RuntimeError("No GPU found. A GPU is needed to quantize model.")
+
+        model.eval()
+
+        # For Transformer model
+        has_config = False
+        if hasattr(model, "config"):
+            has_config = True
+        if hasattr(model, "hf_device_map"):
+            # If the model has a device_map, we don't move to model. We have already dispatch the hook that will do the work
+            has_device_map = True
+        device = get_device(model)
+
+        if has_config:
+            use_cache = model.config.use_cache
+            model.config.use_cache = False
+
+        if self.model_seqlen is None:
+            self.model_seqlen = get_seqlen(model)
+
+        # Step 1: Prepare the data
+        if isinstance(dataset, str):
+            dataset, _ = get_dataset(dataset, tokenizer, seqlen=self.model_seqlen)
+        elif isinstance(dataset, list):
+            dataset = [tokenizer(data, return_tensors="pt") for data in dataset]
+        dataset = prepare_dataset(dataset, pad_token_id=pad_token_id, batch_size=batch_size)
+
+        # Step 2: get the input of the 1st block
+        layer_inputs = []
+        layer_outputs = []
+        layer_input_kwargs = []
+
+        class Catcher(nn.Module):
+            """hijack layer's forward pass to cache data"""
+
+            def __init__(self, m):
+                super().__init__()
+                self.module = m
+
+            def forward(self, input=None, **kwargs):
+                # specific to transformers
+                # some models use all key-value arguments in forward pass call
+                if input is None:
+                    if "hidden_states" in kwargs:
+                        input = kwargs["hidden_states"]
+                    else:
+                        raise ValueError("No input value found in the foward pass")
+                layer_inputs.append(input)
+                other_kwargs = {}
+                for k, v in kwargs.items():  # make sure other arguments also be captured
+                    if k not in ["hidden_states"]:
+                        other_kwargs[k] = v
+                layer_input_kwargs.append(other_kwargs)
+                raise ValueError
+
+        # get block_name
+        if self.block_name_to_quantize is None:
+            self.block_name_to_quantize = get_block_name(model)
+
+        # get modules_name that are preceding the first block
+        if self.module_name_preceding_first_block is None:
+            self.module_name_preceding_first_block = get_preceding_modules(model, self.block_name_to_quantize)
+
+        # get block
+        blocks = get_module_by_name_prefix(model, self.block_name_to_quantize)
+
+        if not has_device_map:
+            # put modules from module_name_preceding_first_block on cuda
```

Comment:
This is already explicit from `module = module.to(0)`, but you could explain why do that rather than a `model = model.to("cuda")`. My guess is that if the user does not use accelerate, we try to avoid OOM by putting only part of the model on cuda device.

Reply:
Yes that's the reason.
 In the case we don't use accelerate, there is two cases: 
- model is on cpu : we only put part of the model on cuda device for inference
- model is on gpu : we do nothing except at the end where we need to put the mode on cpu for packing 
If we use accelerate, we don't need to move the module by manually except at the end where we need to put the mode on cpu for packing.